docs/security/gpg-key-for-ssh-authentication.md

```diff
@@ -0,0 +1,477 @@
+    ssb  4096R/693C5635  created: 2016-04-09  expires: never     
+    ssb  4096R/48B9C23C  created: 2016-04-09  expires: never  
+    (1) Your Name <you@yoursite.net>
+    
+    gpg> key 1
+    
+    sec  4096R/71735D23  created: 2016-04-09  expires: 2017-04-09
+    ssb* 4096R/693C5635  created: 2016-04-09  expires: never     
+    ssb  4096R/48B9C23C  created: 2016-04-09  expires: never
+    (1) Your Name <you@yoursite.net>
+    
+    gpg> keytocard
+    Signature key ....: none
+    Encryption key....: none
+    Authentication key: none
+    
+    Please select where to store the key:
+       (3) Authentication key
+    Your selection? 3
+    
+    You need a passphrase to unlock the secret key for
+    user: "Your Name <you@yoursite.net>"
+    4096-bit RSA key, ID 71735D23, created 2016-04-09
+    
+    gpg> save
+
+This should mean you're ready to roll. If you weren't using a VM/offline machine, then you'll need to delete your computer's copies of the private keys (after safely backing them up), and ensure that the rest of the keys are still on the card.
+
+## Serve your GPG key instead of an SSH key
+
+At this point you should return to your main (still local) machine and import all of the appropriate GPG keys and insert the appropriate GPG device. Install GPG if you need to.
+
+1. Edit the `~/.profile` file (or any similar shell startup file that you choose) to include:
+
+    On Linux:
+
+    {: .file-excerpt}
+    ~/.profile
+    :   ~~~
+        if [ -f "${HOME}/.gpg-agent-info" ]; then
+             . "${HOME}/.gpg-agent-info"
+               export GPG_AGENT_INFO
+               export SSH_AUTH_SOCK
+               export SSH_AGENT_PID
+        fi
+        ~~~
+    
+    On OS X:
+    
+    {: .file-excerpt}
+    ~/.profile
+    :   ~~~
+        [ -f ~/.gpg-agent-info ] && source ~/.gpg-agent-info
+        if [ -S "${GPG_AGENT_INFO%%:*}" ]; then
+            export GPG_AGENT_INFO
+            export SSH_AUTH_SOCK
+            export SSH_AGENT_PID
+        else
+            eval $( gpg-agent --daemon --write-env-file ~/.gpg-agent-info )
+        fi
+        ~~~
+    
+    This ensures that SSH can 'see' your GPG keys, and on OS X, automatically starts `gpg-agent` as needed. Newer versions of GPG may automatically run `gpg-agent`, so check this first before proceeding.
+    
+2. Edit or create `~/.gnupg/gpg-agent.conf`:
+    
+    {: .file-excerpt}
+    ~/.profile
+    :   ~~~
+        default-cache-ttl 600
+        max-cache-ttl 7200
+        enable-ssh-support
+        write-env-file
+        ~~~
+        
+    If you're on OS X and installed GPGTools, you can also add
+
+        pinentry-program /usr/local/MacGPG2/libexec/pinentry-mac.app/Contents/MacOS/pinentry-mac
+    
+    To use the PIN entry program provided by GPGTools.
+    
+3. Restart the GPG agent:
+
+        sudo killall gpg-agent
+
+    gpg-agent --daemon --write-env-file ~/.gpg-agent-info
+
+4. Quit and reopen your shell.
```

Comment:
I chose `reset` here, since I'm not familiar with `source`. The `bash_profile` script needs to be re-run here, which `reset` does.